src/billing_handler.rs

```diff
@@ -0,0 +1,83 @@
+use crate::model::AppState;
+
+use actix_web::web::{Data, Json};
+use actix_web::{get, post, HttpResponse, Responder};
+use anyhow::Context;
+use k256::elliptic_curve::generic_array::sequence::Lengthen;
+use serde_json::json;
+use tiny_keccak::{Hasher, Keccak};
+
+#[derive(Debug, serde::Deserialize)]
+pub struct SigningData {
+    nonce: String,
+    txhashes: Vec<String>,
+}
+
+#[get("/billing/inspect")]
+pub async fn get_bill(appstate: Data<AppState>) -> impl Responder {
+    HttpResponse::Ok().json(json!({
+        "bill": appstate.execution_costs.lock().unwrap().clone(),
+    }))
+}
+
+#[post("/billing/export")]
+pub async fn sign_data(appstate: Data<AppState>, data: Json<SigningData>) -> impl Responder {
+    let signing_data = data.into_inner();
+    if signing_data.nonce.is_empty() {
+        return HttpResponse::BadRequest().body("Nonce must not be empty");
+    }
+
+    if signing_data.txhashes.is_empty() {
+        return HttpResponse::BadRequest().body("List of tx hashes must not be empty");
+    }
+
+    let mut bytes32_nonce = [0u8; 32];
+    if let Err(err) = hex::decode_to_slice(&signing_data.nonce, &mut bytes32_nonce) {
+        return HttpResponse::BadRequest()
+            .body(format!("Error decoding nonce into 32 bytes: {}", err));
+    }
+
+    let mut signed_data: Vec<u8> = bytes32_nonce.to_vec();
+    for txhash in signing_data.txhashes {
+        if let Some(cost) = appstate.execution_costs.lock().unwrap().remove(&txhash) {
+            let mut bytes32_txhash = [0u8; 32];
+            if let Err(err) = hex::decode_to_slice(&txhash[2..], &mut bytes32_txhash) {
+                return HttpResponse::InternalServerError().body(format!(
+                    "Error decoding transaction hash into 32 bytes: {}",
+                    err
+                ));
+            }
+
+            signed_data.append(&mut bytes32_txhash.to_vec());
+            signed_data.append(&mut cost.to_be_bytes().to_vec());
+        } else {
+            return HttpResponse::BadRequest().body(format!(
+                "{} tx hash doesn't exist in the current bill",
+                txhash
+            ));
+        }
+    }
+
+    let mut hasher = Keccak::v256();
+    hasher.update(&signed_data);
+
+    let mut hash = [0u8; 32];
+    hasher.finalize(&mut hash);
+
+    let sign = appstate
+        .signer
+        .sign_prehash_recoverable(&hash)
+        .context("Failed to sign billing data");
+    if sign.is_err() {
+        return HttpResponse::InternalServerError().body(format!("{}", sign.unwrap_err()));
+    }
+    let (rs, v) = sign.unwrap();
+
+    let signature = hex::encode(rs.to_bytes().append(27 + v.to_byte()).as_slice());
+    let signed_data = hex::encode(signed_data.as_slice());
+
+    HttpResponse::Ok().json(json!({
```

Comment:
Is it safe to assume that nonce is supposed to be unique? If yes, then can we save this data with given nonce as key. So that if in a case, client loses the bill, it can query bill again by providing the nonce. And maybe delete the cache after some timeout.

To avoid the case, where client got disconnected from enclave.

Reply:
Yes, the contract will deduplicate receipts based on the nonce

Storing all, or even some, of them is probably overkill, can maybe just store the very last one that the client can then requery (`/billing/latest`). Assuming proper persistence (and API usage) is the responsibility of the client, the only real loss vector that remains is network issues while the response is being transmitted, just the last one should be fine to handle this.

Reply:
We can also protect against duplicate requests by checking the nonce and seeing if it has changed since the last one :thinking:

Reply:
Agree. Nonce check should also be enforced. Also storing the last bill is sufficient to handle the network issue.